pages/science/reactors/extensiblereactor.rst

```diff
@@ -0,0 +1,160 @@
+
+.. math::
+
+   \frac{dm}{dt} = \sum_{in} \dot{m}_{in} - \sum_{out} \dot{m}_{out} +
+                    \dot{m}_{wall}
+                    \tag{1}
+
+.. math::
+
+   m c_p \frac{dT}{dt} = - \dot{Q} - \sum_k h_k \dot{m}_{k,gen}
+       + \sum_{in} \dot{m}_{in} \left(h_{in} - \sum_k h_k Y_{k,in} \right)
+  \tag{2}
+
+.. math::
+
+   m \frac{dY_k}{dt} = \sum_{in} \dot{m}_{in} (Y_{k,in} - Y_k)+
+                      \dot{m}_{k,gen} - Y_k \dot{m}_{wall}
+                      \tag{3}
+
+There are two "sides" to each of these equations: the terms left of the equals
+sign and the terms to the right of the equals sign. This is the format
+in which the user will be editing the governing equations. For example,
+if the user wishes to add a term for a large mass (say a rock) inside
+the reactor to see the effects on reaction temperature:
+
+.. math::
+
+   m c_p \frac{dT}{dt} = - \dot{Q} - \sum_k h_k \dot{m}_{k,gen}
+       + \sum_{in} \dot{m}_{in} \left(h_{in} - \sum_k h_k Y_{k,in} \right)
+  \tag{2}
+
+Will change to:
+
+.. math::
+
+   m c_p \frac{dT}{dt} + m_{rock} c_{p,rock} \frac{dT}{dt} = - \dot{Q} - \sum_k h_k \dot{m}_{k,gen}
+       + \sum_{in} \dot{m}_{in} \left(h_{in} - \sum_k h_k Y_{k,in} \right)
+
+A simple example is shown below to illustrate the process for implementing
+changes in Cantera's existing governing equations.
+We will be replacing the right-hand side (RHS) and left-hand side (LHS) of 
+the temperature governing equation for an Ideal Gas Constant Pressure Reactor.
+All other governing equations defining an Ideal Gas Constant Pressure Reactor
+will remain as the default.
+
+In this example
+
+.. math::
+
+   m c_p \frac{dT}{dt} = - \dot{Q} - \sum_k h_k \dot{m}_{k,gen}
+       + \sum_{in} \dot{m}_{in} \left(h_{in} - \sum_k h_k Y_{k,in} \right)
+  \tag{2}
+
+Will change to:
+
+.. math::
+
+   m_{rock} c_{p,rock}\frac{dT}{dt} + m_{gas}\frac{dT}{dt} = - \dot{Q}
+
+1. Define objects, properties, and initial conditions.
+
+.. code-block:: python
+
+   #create gas object
+   gas = ct.Solution('h2o2.yaml')
+   gas.TPX = 500, ct.one_atm, 'H2:2,O2:1,N2:4'
+   gas_initial_enthalpy = gas.enthalpy_mass
+
+   #define properties of gas and solid
+   mass_gas = 20 #[kg]
+   Q = 100 #[J/s]
+   mass_lump = 10 #[kg]
+   cp_lump = 1.0 #[J/kgK]
+
+   #initialize time at zero
+   time = 0 #[s]
+   n_steps = 300
+
+2. Define a new custom Reactor class. Here we named it "DummyReactor" and 
+chose the Ideal Gas Constant Pressure Reactor as the base class to inheret
+governing equations from. An Extensible Reactor is also known as a Delegated 
+Reactor.
+
+.. code-block:: python
+
+   #define a class representing reactor with a solid mass and gas inside of it
+
+   class DummyReactor(ct.DelegatedIdealGasConstPressureReactor):
+
+      #modify energy equation to include solid mass in reactor
+
+      def after_eval(self, t, LHS, RHS):
+
+         self.m_mass = mass_gas
+
+         LHS[1] = mass_rock*cp_rock+self.m_mass*self.thermo.cp_mass
```

Comment:
```suggestion
         LHS[1] = mass_rock * cp_rock + self.m_mass * self.thermo.cp_mass
```
`mass_rock`, `cp_rock`, `Q`, and `mass_gas` are all undefined, you'll get a `NameError` if you actually try to run this code.

Reply:
I have those variables defined previously in the example, here's the full script the documentatin is based on that I was trying to document.
`    def test_RHS_LHS(self):
        #create gas object
        gas = ct.Solution('h2o2.yaml')
        gas.TPX = 500, ct.one_atm, 'H2:2,O2:1,N2:4'
        gas_initial_enthalpy = gas.enthalpy_mass

        #define properties of gas and solid
        mass_gas = 20 #[kg]
        Q = 100 #[J/s]
        mass_lump = 10 #[kg]
        cp_lump = 1.0 #[J/kgK]

        #initialize time at zero
        time = 0 #[s]
        n_steps = 300

        #define a class representing reactor with a solid mass and
        #gas inside of it
        class DummyReactor(ct.DelegatedIdealGasConstPressureReactor):
            #modify energy equation to include solid mass in reactor
            def after_eval(self,t,LHS,RHS):
                self.m_mass = mass_gas
                LHS[1] = mass_lump*cp_lump+self.m_mass*self.thermo.cp_mass
                RHS[1] = Q

        r1 = DummyReactor(gas)
        r1_net = ct.ReactorNet([r1])

        for n in range(n_steps):
            time += 4.e-4
            r1_net.advance(time)`

Reply:
Apologies if I missed it, but I don't see those variables anywhere else on the page? I'm also not a big fan of splitting a script over multiple code blocks... It's harder to follow as a reader, especially if you're only skimming, and I can't easily copy-paste the parts of the example that might be relevant for my problem, since I have to replace these variables or remove them or rename them

Reply:
Ah I see, there was a change in naming that didn't get updated. `mass_lump` and `lump_cp` are `rock_mass` and `rock_cp` defined on lines 112 and 113. `Q` and `mass_gas` are defined on line 110 and 111. I agree about splitting code, maybe I'll just include the steps as comments in a single code block instead.